### PR TITLE
Support meta scan query in FE

### PR DESCRIPTION
**Pull request by kangkaisen**
For https://github.com/StarRocks/starrocks/issues/559

1. Add meta scan node and operator
2. Add  dict_merge function
3. Add PushDownAggToMetaScanRule

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalMetaScanOperator.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.OperatorVisitor;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class LogicalMetaScanOperator extends LogicalScanOperator {
+    private final Map<Integer, String> aggColumnIdToNames;
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalMetaScanOperator.java]: Fixed

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Statistics.java [hunk 2]
```diff
@@ -32,7 +31,10 @@ public double getOutputSize() {
 
     public ColumnStatistic getColumnStatistic(ColumnRefOperator column) {
         ColumnStatistic result = columnStatistics.get(column);
-        Preconditions.checkState(result != null);
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/Statistics.java]: Fixed, keep this Preconditions check

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+// For meta scan query: select max(a), min(a), dict_merge(a) from test_all_type [_META_]
+// we need to push max, min, dict_merge aggregate function infos to meta scan node
+// we will generate new columns: max_a, min_a, dict_merge_a, make meta scan known what meta info to collect
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java:32]: We add aggregate function name before user-defined columns, So there won't be any confliction.

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java [hunk 4]
```diff
@@ -327,6 +329,20 @@ public Void visitPhysicalSchemaScan(PhysicalSchemaScanOperator node, ExpressionC
...
+    @Override
+    public Void visitLogicalMetaScan(LogicalMetaScanOperator node, ExpressionContext context) {
+        Statistics.Builder builder = estimateScanColumns(node.getTable(), node.getColRefToColumnMetaMap());
+        builder.setOutputRowCount(node.getAggColumnIdToNames().size());
+        return visitOperator(node, context, builder);
+    }
+
+    @Override
+    public Void visitPhysicalMetaScan(PhysicalMetaScanOperator node, ExpressionContext context) {
+        Statistics.Builder builder = estimateScanColumns(node.getTable(), node.getColumnRefMap());
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java:341]:
```suggestion
        Statistics.Builder builder = estimateScanColumns(node.getTable(), node.getColRefToColumnMetaMap());
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java [hunk 5]
```diff
@@ -327,6 +329,20 @@ public Void visitPhysicalSchemaScan(PhysicalSchemaScanOperator node, ExpressionC
         return visitOperator(node, context, builder);
     }
 
+    @Override
+    public Void visitLogicalMetaScan(LogicalMetaScanOperator node, ExpressionContext context) {
+        Statistics.Builder builder = estimateScanColumns(node.getTable(), node.getColRefToColumnMetaMap());
+        builder.setOutputRowCount(node.getAggColumnIdToNames().size());
```
[the remark this answers is not in the document]
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java:335]: The result row count of meta_scan_node row count needn't to accurate，for meta scan node, we could always use two phase aggregate.

**Review by Youngwb — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]

**Pull request merged**